#5588: gh post commit job runs unevenly on ci runners

ttmchiou opened the pull request:
Rebalanced workflows to run in a more readable manner and redistributed fast-dispatch unit tests so that workflow runs faster.
An additional 3 more runners were be required to cut down on post-commit run time for fast-dispatch
Max time reduced down to ~1 hour

Additional Intent:
- Split out the job from 1 step into multiple steps. This will let devs have a more intuitive view of which file generated the crash. File logs will probably be still long but there are technically more points to check the log.
- Devs should still be able to run normally without being affected from these changes except for maybe needing to be aware of how to add new pytests for eager tests? 
- Took advantage of the fact that this will be always fast dispatch so tests are individually called into of using scripts since it was a little messy to split here.

Note:
These changes will indirectly trigger more issues in how we execute workflows as it bypasses calling run_scripts directly. though functionality is still preserved if developers wish to call it on cli on their local machines.
Open to thoughts/opinions on this.

Workflow Example (only fast dispatch)
https://github.com/tenstorrent-metal/tt-metal/actions/runs/8018803993

Full Post-commit run:
https://github.com/tenstorrent-metal/tt-metal/actions/runs/8018807314

ttmchiou committed:
#5588: split pytest scripts in run_python_api_unit_tests into scripts

ttmchiou committed:
#5588: put build docs into script from fast dispatch post commit script

ttmchiou committed:
#5588: redistribute fast dispatch build and unit test workflow jobs

We shorten some names for readability on GH Workflow Web UI.
Split the jobs into smaller jobs for readability purposes too

ttmchiou commented:
draft status pending until post-commit runs pass successfully

ttmchiou committed:
#5588: added sweep test script and call scripts in pipeline instead of directly calling pytests

tt-rkim reviewed (commented):
- Make us codeowners of `tests/scripts/run_pre_post_commit_regressions_fast_dispatch.sh` in `/CODEOWNERS` 
- Add docs in `CONTRIBUTING.md`
  - Change "Running tests in tt-metal" to "Tests in tt-metal"
  - Add new section about adding post commit tests

we're gucci

TT-billteng commented on .github/workflows/fast-dispatch-build-and-unit-tests.yaml:
```diff
@@ -34,8 +35,76 @@ jobs:
+          cd $TT_METAL_HOME
+          export PYTHONPATH=$TT_METAL_HOME
+          ./build/test/tt_metal/unit_tests_fast_dispatch
+      - name: Build Docs
```
is this basically tested [here](https://github.com/tenstorrent-metal/tt-metal/actions/workflows/docs-latest-public.yaml) ?

tt-rkim replied:
@ttmchiou @TT-billteng 
Yes, it is tested in that github runenr workflow, but we didn't add it to the all post commits workflow.

I think there's some sussy stuff about it running on branches, but shouldn't be too hard to fix...

If we want to merge this soon, we could make a separate issue to fix the docs pipeline on branch, delete this step, and then add the docs pipeline to all post commit workflow.

ttmchiou committed:
#5588: Added documentation for split pipelines and guidelines

#5588: Adjusted timeouts on workflow steps for FD
#5588: Addressing PR Comment, adding posix newline at EOF
#5588: Added section on guideline for devs on how to post-commit tests
#5588: Added codeowners to fast_pre_post-commit_regression_fast_dispatch.sh

ttmchiou commented:
@TT-billteng 
@tt-rkim 
Made changes according to comments,
Left the build docs step in as we do want to make sure we keep building docs during post-commit runs.

ttmchiou commented:
Merging.
Completed run here
https://github.com/tenstorrent-metal/tt-metal/actions/runs/8072598531/job/22057660328
